In [1]:
import pandas as pd
import pandas as pd
from sqlalchemy import create_engine
%load_ext sql

In [2]:
# Create a connection to the PostgreSQL database
engine=create_engine('postgresql+psycopg2://postgres:password@localhost:5432/medika_sejahtera')
connection=engine.connect()
print("Connected:", connection.closed == False)

Connected: True


In [3]:
# check connection using SQL magic command
%sql postgresql+psycopg2://postgres:password@localhost:5432/medika_sejahtera

'Connected: postgres@medika_sejahtera'

## 1. Understand Table Structure

*Checking the available columns and their data types*

In [4]:
%%sql

SELECT column_name, data_type 
FROM information_schema.columns 
WHERE table_name = 'dn_data';

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
7 rows affected.


column_name,data_type
delivery_date,date
dn_return_date,date
customer_id,character varying
dn_id,character varying
delivery_method,character varying
status_dn,character varying
invoice_id,character varying


In [5]:
%%sql

SELECT column_name, data_type 
FROM information_schema.columns 
WHERE table_name = 'master_customer';

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
7 rows affected.


column_name,data_type
customer_id,character varying
customer_name,character varying
customer_type,character varying
region,character varying
alamat,character varying
payment_terms,character varying
customer_status,character varying


**Findings**
- `delivery_date` and `dn_return_date` → typed as DATE, meaning date difference calculations can be performed directly.
- `customer_id` is consistent across both tables.

**Conclusion:**  
Table structure is correct, data types are appropriate and all required columns are present.

## 2. Data Preview

*Checking sample records to see if the data makes sense*

In [6]:
%%sql

SELECT *
FROM
    master_customer
LIMIT 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
10 rows affected.


customer_id,customer_name,customer_type,region,alamat,payment_terms,customer_status
RSP001,RSUD Zulaika,RS Pemerintah,Jakarta Barat,"Jl. KH Amin Jasuta No. 173, Cirebon, JA 74749",NET 45,Active
RSS001,RS Suryono,RS Swasta,Tangerang,"Gang Dipatiukur No. 1, Tebingtinggi, Kepulauan Riau 46986",NET 30,Active
KLN001,Klinik Laksita,Klinik,Jakarta Timur,"Gang Surapati No. 0, Payakumbuh, Bali 83302",NET 30,Active
RSS002,RS Wasita,RS Swasta,Jakarta Utara,"Gg. Surapati No. 10, Surakarta, Maluku Utara 15272",NET 30,Active
KLN002,Klinik Sihombing,Klinik,Jakarta Selatan,"Jalan Rajawali Timur No. 088, Palangkaraya, BA 39649",NET 30,Active
RSS003,RS Dongoran,RS Swasta,Bekasi,"Jl. Jend. Sudirman No. 6, Dumai, Sumatera Barat 48478",NET 30,Active
RSP002,RSUD Maheswara,RS Pemerintah,Depok,"Jl. Pelajar Pejuang No. 48, Bandung, DKI Jakarta 56117",NET 45,Active
KLN003,Klinik Wasita,Klinik,Bekasi,"Jalan Abdul Muis No. 1, Sawahlunto, NB 33992",NET 30,Active
KLN004,Klinik Wibisono,Klinik,Jakarta Pusat,"Jalan PHH. Mustofa No. 7, Tomohon, GO 67022",NET 30,Active
KLN005,Klinik Kuswandari,Klinik,Bogor,"Jl. Jend. Sudirman No. 49, Tanjungpinang, KR 36878",NET 30,Active


In [7]:
%%sql

SELECT *
FROM
    dn_data
LIMIT 10;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
10 rows affected.


dn_id,invoice_id,customer_id,delivery_date,delivery_method,dn_return_date,status_dn
DN240001,009501,RSS011,2024-10-11,Internal,2024-11-06,Late
DN240002,009502,RSS011,2024-10-12,Internal,None,Pending
DN240003,009503,RSS052,2024-09-10,Internal,2024-09-16,Received
DN240004,009504,KLN073,2024-02-15,Third-Party Courier,2024-02-20,Received
DN240005,009505,KLN073,2024-02-15,Internal,2024-02-24,Received
DN240006,009506,RSS050,2024-05-31,Internal,2024-06-06,Received
DN240007,009507,RSS050,2024-06-02,Third-Party Courier,2024-06-08,Received
DN240008,009508,RSS068,2024-09-16,Internal,2024-09-18,Received
DN240009,009509,RSP005,2024-12-12,Third-Party Courier,2024-12-20,Received
DN240010,009510,RSP030,2024-06-21,Internal,2024-06-29,Received


**Findings:**
- master_customer
    - `customer_id` values are consistent
    - All `customer_status` values in this sample are Active — worth checking if any inactive customers exist in the full dataset
    - `payment_terms`: Government hospitals (RS Pemerintah) are on NET 45, while private hospitals (RS Swasta) and clinics use NET 30
- dn_data
    - `status_dn` has three possible values: **Received**, **Pending**, and **Late**
    - DNs with **Pending** status have a `dn_return_date` of *None*
    - `delivery_method` splits into two options: **Internal** and **Third-Party Courier**

**Kesimpulan:**
- Data isi kedua tabel masuk akal dan konsisten
- Ditermukan 3 nilai status DN (**Received, Pending,** dan **Late**) yang akan menjadi dasar analisis.

## 3. Count Total Records

*Checking the total number of rows in each table*

In [ ]:
%%sql

SELECT COUNT(*) as total_records
FROM
    master_customer


 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


count
220


In [9]:
%%sql

SELECT COUNT(*) as total_records
FROM
    dn_data

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


total_records
683


**Findings**
- `master_customer`: 220 registered institutions
- `dn_data`: 683 delivery transactions recorded

**Conclusion**
- The dataset is large enough to support meaningful analysis.
- 683 DN transactions across 220 institutions gives a solid, representative picture.

## 4. Check Missing Values

*Checking for missing values in key columns*

In [ ]:
%%sql

SELECT
    COUNT(*) - COUNT(customer_id) AS missing_customer_id,
    COUNT(*) - COUNT(customer_name) AS missing_customer_name,
    COUNT(*) - COUNT(customer_type) AS missing_customer_type,
    COUNT(*) - COUNT(region) AS missing_region
FROM
    master_customer;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


missing_customer_id,missing_customer_name,missing_customer_type,missing_region
0,0,0,0


In [14]:
%%sql

SELECT
    COUNT(*) - COUNT(dn_id) AS missing_dn_id,
    COUNT(*) - COUNT(customer_id) AS missing_customer_id,
    COUNT(*) - COUNT(delivery_date) AS missing_delivery_date,
    COUNT(*) - COUNT(delivery_method) AS missing_delivery_method,
    COUNT(*) - COUNT(dn_return_date) AS missing_dn_return_date,
    COUNT(*) - COUNT(status_dn) AS missing_status_dn
FROM
    dn_data;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


missing_dn_id,missing_customer_id,missing_delivery_date,missing_delivery_method,missing_dn_return_date,missing_status_dn
0,0,0,0,84,0


In [16]:
%%sql

SELECT status_dn, COUNT(*) as jumlah
FROM dn_data
WHERE dn_return_date IS NULL
GROUP BY status_dn;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


status_dn,jumlah
Pending,84


In [17]:
%%sql

SELECT 
    status_dn,
    COUNT(*) AS jumlah
FROM 
    dn_data
GROUP BY 
    status_dn
ORDER BY 
    jumlah DESC;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
3 rows affected.


status_dn,jumlah
Received,484
Late,115
Pending,84


**Findings:**  
Out of 683 total DN transactions, the status breakdown is:
- Received: 484 transactions (70.9%)
- Late: 115 transactions (16.8%)
- Pending: 84 transactions (12.3%)

DN bermasalah (Late + Pending) berjumlah 199 transaksi 
atau 29.1% dari total — hampir 1 dari 3 DN menghambat 
proses penagihan perusahaan.

**Kesimpulan:**
Masalah keterlambatan DN cukup signifikan, hampir sepertiga dari total transaksi berstatus bermasalah dan berpotensi menghambat cash flow perusahaan.

## 5. Cek Duplikat

*Mengecek data yang tercatat lebih dari sekali*

In [20]:
%%sql

SELECT 
    customer_id, 
    COUNT(*) AS jumlah
FROM 
    master_customer
GROUP BY 
    customer_id
HAVING 
    COUNT(*) > 1;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
0 rows affected.


customer_id,jumlah


In [19]:
%%sql

SELECT 
    dn_id, 
    COUNT(*) AS jumlah
FROM 
    dn_data
GROUP BY 
    dn_id
HAVING 
    COUNT(*) > 1;

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
0 rows affected.


dn_id,jumlah


**Temuan:**
Tidak ditemukan duplikat di dn_data maupun master_customer.
Setiap dn_id dan customer_id bersifat unik.

**Kesimpulan:**
Data bersih dari duplikat — tidak ada transaksi atau 
instansi yang tercatat lebih dari sekali.

## 6. Cek Konsistensi Antar Tabel

*Mengecek apakah customer_id di dn_data semua terdaftar di master_customer*

In [4]:
%%sql

SELECT 
    COUNT(*) AS customer_tidak_terdaftar
FROM 
    dn_data
WHERE 
    customer_id NOT IN 
        (
        SELECT customer_id FROM master_customer
        );

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
1 rows affected.


customer_tidak_terdaftar
0


**Temuan:**
- Semua customer_id di dn_data terdaftar di master_customer.
- Tidak ada customer_id yang putus antar tabel.

**Kesimpulan**
- Semua customer_id di dn_data terdaftar di master_customer.
- Tidak ada customer_id yang putus antar tabel.

## 7. Cek Tipe Data Kritis

*Memastikan kolom tanggal bertipe DATE, bukan teks*

In [5]:
%%sql

SELECT 
    column_name,
    data_type
FROM 
    information_schema.columns
WHERE 
    table_name = 'dn_data'
AND 
    column_name IN ('delivery_date', 'dn_return_date');

 * postgresql+psycopg2://postgres:***@localhost:5432/medika_sejahtera
2 rows affected.


column_name,data_type
delivery_date,date
dn_return_date,date


**Temuan:**
- delivery_date → DATE 
- dn_return_date → DATE 

**Kesimpulan**
Kedua kolom tanggal sudah bertipe DATE — perhitungan selisih hari antara delivery_date dan dn_return_date dapat langsung dilakukan tanpa konversi tambahan